### PR TITLE
Feature/bouncy pixels animation

### PR DESCRIPTION
**Pull request by jonasgrootkormelink**

### CODE REVIEW

File: client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelsAnimation.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import com.cleverfranke.util.PColor;
+import nl.sensorlab.videowall.animation.BaseAnimation;
+import processing.core.PApplet;
+import processing.core.PGraphics;
+
+public class BouncyPixelsAnimation extends BaseAnimation{
+
+	public PApplet parent;
+
+	BouncyPixelSystem  bouncypixelssytem;
```
Comment (wilcotomassen) [inline client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelsAnimation.java]: Fix member visibility here

File: client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelSystem.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package nl.sensorlab.videowall.animation.baseanimations.bouncypixels;
+
+import java.util.ArrayList;
+
+import nl.sensorlab.videowall.animation.BaseAnimation;
+import processing.core.PGraphics;
+
+public class BouncyPixelSystem {
+	
+	BouncyPixelsAnimation parent;
```
Comment (wilcotomassen) [inline client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelSystem.java]: Fix member visibility here

File: client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixel.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+					// Update b (change direction)
+					b.velocity.x += ax;
+					b.velocity.y += ay;
+				}		
+			}
+		}
+	}
+
+	public void update() {
```
Comment (wilcotomassen) [inline client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixel.java]: This method should also use delta time to decouple animation speed and frame rate

File: client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixel.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+import java.util.ArrayList;
+
+import nl.sensorlab.videowall.animation.BaseAnimation;
+import processing.core.PGraphics;
+import processing.core.PVector;
+
+public class BouncyPixel {
+
+	BouncyPixelSystem parent;
```
Comment (wilcotomassen) [inline client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixel.java]: Again: member visibility: parent has no visibility declared, do all other members need to be public?

File: client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixel.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+	public PVector position;
+	public PVector velocity;
+	public float diameter;
```
Comment (wilcotomassen) [inline client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixel.java]: If you work with radius here, it saves you a boatload of diameter / 2 in the update functions. Makes for better performance.

File: client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelSystem.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	public void generateBouncyPixels(int amountBouncyPixels) {
+		for(int i = 0; i < amountBouncyPixels; i++) {
+			addBouncyPixel();
+		}
+	}
+	
+	public void addBouncyPixel() { 
+		float x = (float)(Math.random() * BaseAnimation.PIXEL_RESOLUTION_X);
+		float y = -(float)(Math.random() * BaseAnimation.PIXEL_RESOLUTION_Y);// Start outside window height
+		float diameter = 1; // <-- Because it are pixels we want this to be set to 1
```
Comment (wilcotomassen) [inline client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelSystem.java]: If this is fixed (static) why isn't it a constant? Then it doesn't need to be in memory and not passed around in methods.

File: client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelsAnimation.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
+package nl.sensorlab.videowall.animation.baseanimations.bouncypixels;
+
+import java.util.ArrayList;
+import com.cleverfranke.util.PColor;
+import nl.sensorlab.videowall.animation.BaseAnimation;
+import processing.core.PApplet;
+import processing.core.PGraphics;
+
+public class BouncyPixelsAnimation extends BaseAnimation{
```
Comment (wilcotomassen) [inline client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelsAnimation.java]: This class feels like it's just a wrapper around BouncyPixelSystem. I think you can merge the whole BouncyPixelSystem in this class without losing any clarity.

File: client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelsAnimation.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+		g.noStroke();
+		g.rect(0, 0, PIXEL_RESOLUTION_X, PIXEL_RESOLUTION_Y);
+		
+		bouncypixelssytem.update();
+		bouncypixelssytem.draw(g);
+		
+		if(addBouncyPixelCounterMillis > ADD_BOUNCY_PIXEL_EVERY_MILLIS) {
+			addBouncyPixelCounterMillis = 0;
+			bouncypixelssytem.addBouncyPixel();
+			if(bouncypixelssytem.bouncypixels.size() > MAX_AMOUNT_BOUNCY_PIXELS) {
```
Comment (wilcotomassen) [inline client/src/nl/sensorlab/videowall/animation/baseanimations/bouncypixels/BouncyPixelsAnimation.java]: Wouldn't it make sense to give each BouncyPixel a lifetime, and then remove it once it dies?

**Review by wilcotomassen — CHANGES_REQUESTED** [review]

**Review by wilcotomassen — CHANGES_REQUESTED** [review]
I think you can Merge BouncyPixelSystem into BouncyPixelAnimation and nest BouncyPixel in BouncyPixelAnimation to remove the need for a package

**Comment by jonasgrootkormelink** [comment]
@wilcotomassen I think we can merged this branch into develop. I've made the requested changes!

**Review by wilcotomassen — APPROVED** [review]

**Pull request merged**